content/post/104-dht-deep-dive.md

```diff
@@ -0,0 +1,154 @@
+
+## Kademlia Overview
+
+The Kademlia algorithm has been around for a while and there are some great resources already available for it online, including the [paper itself](https://ipfs.io/ipfs/QmaVrnwZrnoG4YramcN75mbE5AUfCymiEErrHGXoQR968V) and [Wikipedia](https://en.wikipedia.org/wiki/Kademlia). We’ll go through some of the basics here, though.
+
+The general idea of Kademlia is to build a DHT on top of three system parameters:
+
+*   An address space: a way that all of the peers in the network can be uniquely identified (in IPFS this is all the numbers from `0` to `2^256-1`)
+*   A metric to order the peers in the address space and therefore visualize all the peers along a line ordered from smallest to largest (in IPFS we we take `SHA256(PeerID)` and interpret it as an integer between `0` and `2^256-1`)
+*   A projection, that will take a record key and calculate a position in the address space where the peer(s) most ideally suited to store the record should be near (in IPFS we use `SHA256(Record Key)`)
+
+Having this address space and a peer ordering metric allows us to search the network as though it was a sorted list. In particular, we can turn the system into something like a skip list where a peer knows peers that are distances of around 1,2,4,8,16… away from it. This will allow us to search the list in time that is logarithmic in the size of the network (i.e. O(log(N)) lookup time). Unlike a skip list, Kademlia is somewhat unstable since peers can join, leave and rejoin the network at any time. To deal with the unstable nature of the system, a Kademlia peer does not just keep links to the peers with distance 1,2,4,8,... away from it, but instead for each multiple of 2 away keeps up to `K` (in IPFS 20) links. For example, instead of a peer keeping a single link 128 away it would keep 20 links that are between 65 and 128 away.
+
+Note that the selection of network wide parameters like `K` is not arbitrary, but is instead determined based on the observed/expected churn in the network and the frequency with which the network will republish information. These system and network parameters drive the decisions made in Kademlia’s two main components, the routing table which tracks all those links in the network, and the lookup algorithm which determines how to traverse those links to store and retrieve data.
+
+## Kademlia and Undialable Peers
+
+As noted above, a major property of Kademlia is that all peers can be assembled on to a line from smallest to largest. This property is useful since it means that as peer 0 walks down the line to find peer 55 it can know it is getting progressively closer. However, this requires that everyone on the line can talk to each other, otherwise peer 33 might send peer 0 down a dead end by telling them the content they want is on a node they can’t communicate with. This can result in the network being slow, and more importantly, **fragmented** with data being accessible by some peers and not others.
```

Comment:
"on to a line" -> "inline"
```suggestion
As noted above, a major property of Kademlia is that all peers can be assembled inline from smallest to largest. This property is useful since it means that as peer 0 walks down the line to find peer 55 it can know it is getting progressively closer. However, this requires that everyone on the line can talk to each other, otherwise peer 33 might send peer 0 down a dead end by telling them the content they want is on a node they can’t communicate with. This can result in the network being slow, and more importantly, **fragmented** with data being accessible by some peers and not others.
```